runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,481 @@
+      return;
+    }
+
+    // At this point if we have any deferred data, we must pass this to some
+    // other thread, as we could be parked indefinitely. Find first active
+    // thread to dump the rest of our work on...
+    struct tls_node *node = get_tls_list();
+    for (; node != NULL; node = node->next) {
+      if (get_tasks(node) > 0) {
+        // Double check...
+        acquire_spinlock(node);
+        if (get_tasks(node) > 0) {
+          break;
+        }
+        release_spinlock(node);  
+      }
+    }
+
+    // If node is NULL, then all threads are finished so we should be able to safely
+    // handle our deferred data...
+    if (node == NULL) {
+      handle_deferred_data(tls);
+    } else {
+      // If the node is not NULL, we found our thread to dump on and we already hold its lock.
+      struct defer_node *list = pop_all_defer_list(tls);
+      
+      // If their list is empty its easy
+      if (node->deferList == NULL) {
+        node->deferList = list;
+      } else {
+        // TODO: Optimize! If tls->deferList is size N and node->deferList is size M,
+        // then is O((M+N) * N) = O(M*N^2).
+        while (list != NULL) {
+          struct defer_node *dnode = list;
+          list = list->next;
+
+          // Insert as head
+          if (dnode->targetEpoch >= node->deferList->targetEpoch) {
+            dnode->next = node->deferList;
+            node->deferList = dnode;
+            continue;
+          }
+
+          // Insert between nodes...
+          struct defer_node *ddnode = node->deferList, *prev = NULL;
+          for (; ddnode != NULL; ddnode = ddnode->next) {
+            if (ddnode->targetEpoch >= dnode->targetEpoch) {
+              assert(prev != NULL);
+              dnode->next = ddnode->next;
+              ddnode->next = dnode;
+              break;
+            }
+
+            prev = ddnode;
+          }
+
+          // Exhausted all options, make as tail...
+          if (ddnode == NULL) {
+            assert(prev != NULL);
+            prev->next = dnode;
+          }
+        }
+      }
+
+      release_spinlock(node);
+    }
+
+    release_spinlock(tls);
+  } else {
+    // Whenever we destroy a task, handle destroying any deferred data as well.
+    acquire_spinlock(tls);
+    handle_deferred_data(tls);
+    release_spinlock(tls);
+  }
+}
+
+void chpl_qsbr_checkpoint(void) {
+    struct tls_node *tls = CHPL_TLS_GET(chpl_qsbr_tls);
+    if (tls == NULL) {
+        init_tls();
+        tls = CHPL_TLS_GET(chpl_qsbr_tls);
+    }
+
+    // Observe the current epoch.
+    observe_epoch(tls);
+
+    // Check if we have passed enough checkpoints to process our deferred data.
+    if (++tls->passedCheckpoints % CHPL_QSBR_CHECKPOINTS_PER_PROCESS) {
+      acquire_spinlock(tls);
+      handle_deferred_data(tls);
+      release_spinlock(tls);
+    } 
+}
+
+void chpl_qsbr_defer_deletion(void *data) {
+    uint64_t epoch = atomic_fetch_add_uint_least64_t(&global_epoch, 1) + 1;
```

Comment:
Did you mean for this to call advance_global_epoch? Remember to keep the + 1 fix somewhere if you change it!